File: solutions/algorithms/algorithm_selection_sort.js

// create the array to sort
var arr = createArrayToSort(1000, 0, 10000)

// create a sorted version of the array for validation
var validationArr = [].concat(arr).sort(function (a, b){ return a - b })

// sort the array
var sortedArr = selectionSort([].concat(arr))

// compare the native and custom sorts
try {
  compareArrays(validationArr, sortedArr)
  console.log("THE ARRAYS MATCHED!")
} catch (e) {
  console.error(e.message)
}

/**
 * Sort an array w/ selection sort
 * @param  {Array.<number>} sortArr the array to sort
 * @return {Array.<number>} returns the array to be sorted
 */
function selectionSort(sortArr) {
  for (var i = 0; i < sortArr.length - 1; i++) {
    var minIdx = i, minVal = sortArr[i]

    for (var j = i + 1; j < sortArr.length; j++) {
      if (sortArr[j] < minVal) {
        minIdx = j
        minVal = sortArr[j]
      }
    }

    if (minIdx != i) {
      var temp = sortArr[i]
      sortArr[i] = sortArr[minIdx]
      sortArr[minIdx] = temp
    }
  }

  return sortArr
}

/**
 * Compare 2 arrays and make sure each element matches
 *
 * @param  {Array.<number>} arr1 the first array
 * @param  {Array.<number>} arr2 the second array
 */
function compareArrays(arr1, arr2) {
  if (arr1.length != arr2.length) throw new Error("The array lengths did not match")
  for (var i = 0; i < arr1.length; i++) {
    if (arr1[i] != arr2[i]) throw new Error("The arrays did not match")
  }
}

/**
 * Create an array of numbers to sort
 *
 * @param  {number} length the number of items to insert into the array
 * @param  {number} min    the minimium value for each item
 * @param  {number} max    the maximum value for each item (non-inclusive)
 * @return {Array.<number>} The array of numbers
 */
function createArrayToSort(length, min, max) {
  var rangeSize = max - min
  var items = []
  for (var i = 0; i < length; i++) {
    items.push(Math.floor(Math.random() * rangeSize) + min)
  }
  return items
}

File: solutions/algorithms/algorithm_selection_sort.py
import random

# Basic selection sort
def selectionSort(arr):
  newArr = arr[:]
  listLen = len(arr)

  for i in range(0, listLen - 1):
    lowestIdx = i
    lowestVal = newArr[i]

    # Find the lowest val
    for j in range(i + 1, listLen):
      if newArr[j] < lowestVal:
        lowestIdx = j
        lowestVal = newArr[j]

    if i != j:
      temp = newArr[i]
      newArr[i] = newArr[lowestIdx]
      newArr[lowestIdx] = temp

  return newArr

# Create an array to sort
def createArrayToSort(len, min, max):
  arr = []
  for i in range(0, len):
    arr.append(random.randint(min, max))
  return arr

# Compare two arrays and make sure they match
def compareArrays(arr1, arr2):
  len1 = len(arr1)
  len2 = len(arr2)
  if len1 != len2:
    print "The list lengths did not match"
    return False
  for i in range(0, len1):
    if arr1[i] != arr2[i]:
      print str.format("The lists did not match at index {0}", i)
      return False
  return True

# Create the test array
arr = createArrayToSort(1000, 0, 10000)

# Clone and sort the test array using the built in functionality
validationArr = sorted(arr)

# Sort the array using the custom sort
sortedArr = selectionSort(arr)

# Verify that the arrays match
listsMatch = compareArrays(validationArr, sortedArr)
if listsMatch:
  print "THE LISTS MATCH"